src/hoisted.jl

```diff
@@ -0,0 +1,91 @@
+# Specialized functions for increased performance when JSON is in-memory
+
+addbyte!(x::Int, c) = x + 1
+addbyte!(x::Vector{UInt8}, c) = push!(x, c)
+
+function parse_string(ps::MemoryParserState)
+    # "Dry Run": find length of string so we can allocate the right amount of
+    # memory from the start.
+    s = ps.s
+    len = parse_string(ps, 0)
+
+    # Now read the string itself
+    ps.s = s
+    @static if Compat.UTF8String == Base.String
+        b = sizehint!(UInt8[], len)
+        parse_string(ps, b)
+        ccall(:jl_array_to_string, Ref{String}, (Any,), b)
+    else
+        ret = String(UInt8[])
+        b = sizehint!(ret.data, len)
+        parse_string(ps, b)
+        ret
+    end
```

Comment:
It doesn't seem worthwhile to have these two cases.   Can we just use the second case, with `ret = Compat.UTF8String(UInt8[])`, in all Julia versions?

In the longer run, there should probably be an `unsafe_string` added to Base, and then to Compat.